[23.0] Skip installing npm/yarn if available, fix conditional dependency parsing, create virtualenv via conda when conda active

bernt-matthias opened the pull request:
I ran into the following problem (when upgrading to 23.0) with node installed via pip

```
node: /lib64/libm.so.6: version `GLIBC_2.27' not found (required by node)
node: /lib64/libc.so.6: version `GLIBC_2.25' not found (required by node)
node: /lib64/libc.so.6: version `GLIBC_2.28' not found (required by node)
node: /lib64/libstdc++.so.6: version `CXXABI_1.3.9' not found (required by node)
node: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.20' not found (required by node)
node: /lib64/libstdc++.so.6: version `GLIBCXX_3.4.21' not found (required by node)
```

I tried to circumvent this by using node installed via conda (which seemed to be supported in the startup script). But 
apparently, the script never was able to install node (and yarn) via conda (since the python venv is always activated and `$VIRTUAL_ENV` is set).

Not sure about  enforcing the minimal Python version (instead of enforcing the minimal Python version exactly) ... 

I'm fine if this should target dev (I can cherry-pick this on my release branch meanwhile).

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [x] This is a refactoring of components with existing test coverage.
  - **I hope/assume**  ..  I'm currently running this on my testing instance
- [ ] Instructions for manual testing are as follows:
  
## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

nsoranzo committed:
Test first startup on macOS also with Python 3.11

- Allow passing GALAXY_CONDA_PYTHON_VERSION environment variable to
  the startup script to specify the Python version to use when
  creating the `_galaxy_` conda environment.
- Some Shellcheck fixes.

nsoranzo committed:
Pin pysam to >=0.21.0

This reverts commit 353c08cbb942a766e1c1a5e96060d2032bf2a1a0 .

pysam 0.20.0 doesn't build for Python 3.11 on macOS, so we need to
temporarily add the unnecessary Cython runtime dependency that
pysam 0.21 brings in.

nsoranzo committed:
Split matplotlib requirement to support Python 3.11 on macOS

matplotlib 3.5.3 (the latest version supporting Python 3.7) fails to
build a wheel for Python 3.11 on macOS. Although it can be still be
installed via the `setup.py install` fallback, this has been removed
in pip 23.1 .

bernt-matthias committed:
Improve log message wording

nsoranzo commented on scripts/common_startup.sh:
```diff
@@ -241,16 +241,16 @@ fi
+    if [ -n "$CONDA_DEFAULT_ENV" ] && [ -n "$CONDA_EXE" ]; then
         if ! in_conda_env "$(command -v yarn)"; then
             echo "Installing yarn into '$CONDA_DEFAULT_ENV' Conda environment with conda."
             $CONDA_EXE install --yes --override-channels --channel conda-forge --name "$CONDA_DEFAULT_ENV" yarn
```
The macOS build, which uses conda, fails with:
```
Unknown Syntax Error: Unsupported option name ("--check-files").
```
which it's due to the fact that yarn 3 is installed instead of yarn "classic".

```suggestion
            $CONDA_EXE install --yes --override-channels --channel conda-forge --name "$CONDA_DEFAULT_ENV" 'yarn<2'
```

nsoranzo commented on lib/galaxy/dependencies/pinned-requirements.txt:
```diff
@@ -137,7 +137,8 @@ pyparsing==3.0.9 ; python_version >= "3.7" and python_version < "3.12"
 pyrsistent==0.19.3 ; python_version >= "3.7" and python_version < "3.12"
-pysam==0.20.0 ; python_version >= "3.7" and python_version < "3.12"
+pysam==0.20.0 ; python_version >= "3.7" and python_version < "3.11"
+pysam==0.21.0 ; python_version >= "3.11" and python_version < "3.12"
```
Can we just use pysam==0.21.0 for all Python versions? That's what we have on 23.1 and dev.

bernt-matthias replied:
Sure, wasn't aware of this. Thanks for the hint.

bernt-matthias committed:
fix job config parsing for getting conditional requirements

noticed that I did not get `drmaa` installed automatically.
reason is that self.config.get("job_config_file") only gets
the name of the config file relative to the path containing
the galaxy config.

bernt-matthias committed:
activate the conda env before the python venv

and install node and yarn via conda if a conda env is used

bernt-matthias committed:
also set galaxy config file variable in common startup

otherwise GALAXY_CONDITIONAL_DEPENDENCIES can not be determined.
in run.sh set_galaxy_config_file_var is called only after common_startup
is called.

nsoranzo commented on scripts/common_startup.sh:
```diff
@@ -109,15 +109,15 @@ if [ $SET_VENV -eq 1 ] && [ $CREATE_VENV -eq 1 ]; then
                     echo "To avoid this, use the --no-create-venv flag or set \$GALAXY_CONDA_ENV to an"
                     echo "existing environment before starting Galaxy."
-                    $CONDA_EXE create --yes --override-channels --channel conda-forge --name "$GALAXY_CONDA_ENV" "python=${MIN_PYTHON_VERSION}" "pip>=${MIN_PIP_VERSION}"
+                    $CONDA_EXE create --yes --override-channels --channel conda-forge --name "$GALAXY_CONDA_ENV" "python>=${MIN_PYTHON_VERSION}" "pip>=${MIN_PIP_VERSION}"
```
This could install a Python version that we don't support yet (e.g. the upcoming 3.12), is it necessary?

bernt-matthias replied:
I just followed the variable name here: `MIN_PYTHON_VERSION`. It's not necessary. Maybe we can reuse this: 
https://github.com/galaxyproject/galaxy/blob/f2fcfa60f9e22f21179118e3f149ee01c02deb67/lib/galaxy/dependencies/update.sh#L10C41-L10C41?

The [docs](https://docs.galaxyproject.org/en/master/admin/python.html)  currently just state that Python >=3.7 is supported. Maybe we want to be more specific there? But I was also wondering if we should extend our testing here. As far as I get it we only test 3.7, or?

nsoranzo replied:
> The [docs](https://docs.galaxyproject.org/en/master/admin/python.html) currently just state that Python >=3.7 is supported. Maybe we want to be more specific there?

Yes, it would be good to specify also the latest Python version we are testing with.

> But I was also wondering if we should extend our testing here. As far as I get it we only test 3.7, or?

```shell
$ grep -r '3\.11' .github/workflows/
.github/workflows/test_galaxy_packages.yaml:        python-version: ['3.7', '3.11']
.github/workflows/first_startup.yaml:        python-version: ['3.7', '3.11']
.github/workflows/reports_startup.yaml:        python-version: ['3.7', '3.11']
.github/workflows/lint_openapi_schema.yml:        python-version: ['3.7', '3.11']
.github/workflows/osx_startup.yaml:        python-version: ['3.7', '3.11']
.github/workflows/lint.yaml:        python-version: ['3.7', '3.11']
.github/workflows/unit.yaml:        python-version: ['3.7', '3.11']
```

where `lint_openapi_schema.yml` and `osx_startup.yaml` where added in 23.1 .

nsoranzo replied:
> I just followed the variable name here: MIN_PYTHON_VERSION. It's not necessary. Maybe we can reuse this:
https://github.com/galaxyproject/galaxy/blob/f2fcfa60f9e22f21179118e3f149ee01c02deb67/lib/galaxy/dependencies/update.sh#L10C41-L10C41?

This was already changed for 23.1 in commit 940ff9fe36af0a1b3ce02851fe2f8103b9dadabb so you can use the `GALAXY_CONDA_PYTHON_VERSION` environment variable to specify which Python version to use when creating the conda environment, would that work for you?

bernt-matthias replied:
Sounds reasonable. Thanks for all the helpful explanations.

nsoranzo replied:
Cool, then I'll remove this change and add a small refactoring to this branch if that's OK?

bernt-matthias replied:
Sure.

nsoranzo committed:
Code refactoring

Simplification allowed by commit 39c0000c1bbf6c96f55f0143bc788f82107745a9 .

nsoranzo committed:
Do not install npm/yarn via conda if they already available at the correct version

dannon reviewed (approved):
Thanks, both @bernt-matthias and @nsoranzo!
